### DISCUSSION TITLE
DDR4 SDRAM - Initialization, Training and Calibration

### DISCUSSION THREAD

**Discussion by subbdue** [subbdue/systemverilog.io#3] [Website Comments]
Please use this Discussions thread for questions and comments on the article `DDR4 SDRAM - Initialization, Training and Calibration`

**Comment by ipp12** [comment]
Hi,

There is a correction in the ZQ calibration part - drive strength is for WRITEs & termination for READs

**Comment by NhatVoDinh** [comment]
what is the value of VOH[0:3] , Is it swing from 0 to VDD? I am not clear about the difference of VOH and PU signals , which is the control sig nal? and which is the data transfer signal?

**Comment by mohsin-code** [comment]
What is CTT and SSTL and POD?
What is push-pull vs pseudo open drain I/O?
What is termination resistance?
Why we calibrate it to VDDq/2? For DDR4, is it VrefDQ now?
Are there 5 p-channel connected to all DQ lines or each has unique.
How many parallel resistors does a pin have?
In write leveling, does controller send DQS at CK to DRAM, and sets delay based on returned CK and DQ (0/1)?
In write leveling, are DQS pulses sample the CK and return it back (What that mean?)
In write leveling, step 1, 2 looks same, 3 sets a delay, 4, 5 and 6 are just repititions and 7 is remove out of mode?
MPR Pattern Write is used before the next stages for the next stages to store and send DATA instead of the banks itself. Does it have all address spaces and data space? Data Patterns, Error log, Mode Register and DRAM manufacture (RD) seem too much for 8-bit registers.
In read centering, is PHY trained or the controller?
In read centering, MR3[2]=1? then it changes the interface to MPR instead of memory core?
In read centering, step 3 and 4 looks like the same thing, and 5 is repeat for all DQ.
If we interface with MPR instead of Mem core, then wouldn't the delay be wrong?
In write centering step 1, does it do continuous stream of W-R-S-C?
